File: apps/api/app/models/protocol_template.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSON
from sqlalchemy.orm import Mapped, mapped_column

from app.database import UUID, Base


class TemplateCategory(str, enum.Enum):
    longevity = "longevity"
    sleep = "sleep"
    cognitive = "cognitive"
    energy = "energy"
    immune = "immune"
    recovery = "recovery"
    hormonal = "hormonal"
    skin = "skin"
    gut = "gut"
    cardiovascular = "cardiovascular"
    starter = "starter"


class ProtocolTemplate(Base):
    """System-provided protocol blueprints that users can browse and adopt."""

    __tablename__ = "protocol_templates"

    id: Mapped[uuid.UUID] = mapped_column(UUID(), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    category: Mapped[str] = mapped_column(String(30), nullable=False, index=True)
    difficulty: Mapped[str | None] = mapped_column(String(20), nullable=True)  # beginner, intermediate, advanced
    icon: Mapped[str | None] = mapped_column(String(30), nullable=True)  # FontAwesome icon name
    is_featured: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)

    # JSON array of item blueprints:
    # [{"type": "supplement", "catalog_name": "Vitamin D3", "dosage": "5000 IU",
    #   "take_window": "morning_with_food", "frequency": "daily", "notes": "..."}]
    items: Mapped[list[dict] | None] = mapped_column(JSON, nullable=True)

    # Tags for filtering/search
    tags: Mapped[list[str] | None] = mapped_column(JSON, nullable=True)

    # Adoption count — tracks popularity
    adoption_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

File: packages/domain/src/protocol-template.ts
export interface TemplateItemBlueprint {
  type: string;
  catalog_name: string;
  dosage?: string | null;
  take_window?: string | null;
  frequency?: string | null;
  notes?: string | null;
}

export interface ProtocolTemplate {
  id: string;
  name: string;
  description: string | null;
  category: string;
  difficulty: string | null;
  icon: string | null;
  is_featured: boolean;
  items: TemplateItemBlueprint[] | null;
  tags: string[] | null;
  adoption_count: number;
  created_at: string;
}

export interface ProtocolTemplateListItem {
  id: string;
  name: string;
  description: string | null;
  category: string;
  difficulty: string | null;
  icon: string | null;
  is_featured: boolean;
  items_count: number;
  tags: string[] | null;
  adoption_count: number;
}

export interface AdoptTemplateResponse {
  protocol_id: string | null;
  protocol_name: string;
  items_created: number;
  items_existing: number;
  message: string;
}

export type TemplateCategory =
  | "longevity"
  | "sleep"
  | "cognitive"
  | "energy"
  | "immune"
  | "recovery"
  | "hormonal"
  | "skin"
  | "gut"
  | "cardiovascular"
  | "starter";
